newpackage: py-awscli (and dependencies)

neilflood opened the pull request:
Package for command line tools for Amazon Web Services. Also 5 other packages it requires which were not already in Spack.  

I have tried to put in all the version restrictions contained in all the various setup.py and requirements.txt files associated with these packages. A few of the single-version requirements seemed overly restrictive, so I have left out a few, but may well be proven wrong in the end. It was a bit of a nightmare to track through......

If anyone thinks the main package should not have the leading "py-" I am open to changing that. I just went with what Spack wanted to call it.

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+
+    version('1.16.111', sha256='63c7900e65a76a1574e1d6e3aaa37c98403339a552428b47559d4e05081f8e4b')
+
+    depends_on('py-setuptools')
```
`type='build'`

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+    version('1.16.111', sha256='63c7900e65a76a1574e1d6e3aaa37c98403339a552428b47559d4e05081f8e4b')
+
+    depends_on('py-setuptools')
+    depends_on('py-docutils@0.10:')
```
Probably `type=('build', 'run')`

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+
+    depends_on('py-setuptools')
+    depends_on('py-docutils@0.10:')
+    depends_on('py-nose', type=('build', 'test'))
```
Probably only `type='test'`

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+    depends_on('py-setuptools')
+    depends_on('py-docutils@0.10:')
+    depends_on('py-nose', type=('build', 'test'))
+    depends_on('py-tox@2.3.1:2.999', type='test')
```
I don't think this dependency is actually required, probably only needed for version compatibility tests in Travis upstream.

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+    depends_on('py-nose', type=('build', 'test'))
+    depends_on('py-tox@2.3.1:2.999', type='test')
+    depends_on('py-colorama@0.2.5:0.3.9', type=('build', 'run'))
+    depends_on('py-mock@1.3.0:', type=('build', 'run'))
```
Probably only `type='test'`

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+    depends_on('py-colorama@0.2.5:0.3.9', type=('build', 'run'))
+    depends_on('py-mock@1.3.0:', type=('build', 'run'))
+    depends_on('py-rsa@3.1.2:3.5.0', type=('build', 'run'))
+    depends_on('py-wheel@0.24.0:', type=('build', 'run'))
```
I don't think this dependency is actually needed to build the package. Probably only if you want to create a wheel.

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+    depends_on('py-rsa@3.1.2:3.5.0', type=('build', 'run'))
+    depends_on('py-wheel@0.24.0:', type=('build', 'run'))
+    depends_on('py-pyyaml@3.10:3.13', type=('build', 'run'))
+    depends_on('py-argparse', when='^python@2.6:2.6.99', type=('build', 'run'))
```
`py-argparse@1.1:`

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+    # I suspect there will need to be some version dependencies for these,
+    # too, but they are not explicit. There is an ominous message about tandem
+    # development in the requirements.txt
+    depends_on('py-jmespath', type=('build', 'run'))
```
This dependency isn't mentioned in `setup.py`, so I'm not sure if it's required.

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+    # too, but they are not explicit. There is an ominous message about tandem
+    # development in the requirements.txt
+    depends_on('py-jmespath', type=('build', 'run'))
+    depends_on('py-s3transfer', type=('build', 'run'))
```
`py-s3transfer@0.2.0:0.2.999'

adamjstewart commented on var/spack/repos/builtin/packages/py-awscli/package.py:
```diff
@@ -0,0 +1,34 @@
+    # development in the requirements.txt
+    depends_on('py-jmespath', type=('build', 'run'))
+    depends_on('py-s3transfer', type=('build', 'run'))
+    depends_on('py-botocore', type=('build', 'run'))
```
`py-botocore@1.12.101`
If you don't give it the exact version, setuptools will try to automatically download and install the correct version, which won't work on an air-gapped network.

neilflood commented:
Thanks @adamjstewart. You have warned me about requirements.txt before, but I never learn. I will write it out 100 times.....

Fixed versions as suggested. It seems you are quite correct that tox is not really necessary, I did not know much about it, but have now removed it. Also realised I had included boto3 without intending to. It was intended to be a separate PR - while it is also a part of Amazon's AWS client code, it is not a requirement for AWS CLI - so I have removed it from this PR.

adamjstewart commented on var/spack/repos/builtin/packages/py-botocore/package.py:
```diff
@@ -0,0 +1,39 @@
+
+    version('1.12.101', sha256='46e4daaa7c8cb29237802b63699c16a116f96f301ad2fcfef800574333b58b98')
+    version('1.12.61', sha256='946c24b616cc885d490a6999026125524e85751540eb4af501673d0f5bc7eee1')
+
```
Missing build dep on setuptools.

neilflood replied:
Done

adamjstewart commented on var/spack/repos/builtin/packages/py-botocore/package.py:
```diff
@@ -0,0 +1,39 @@
+    depends_on('py-urllib3@1.20:1.23', type=('build', 'run'),
+               when='^python@3.3')
+    depends_on('py-urllib3@1.20:1.25', type=('build', 'run'),
+               when='^python@2.7')
```
Can you confirm that versions like `2.7` actually work? I've had a lot of trouble with shortened versions recently (see #8957)

neilflood replied:
Hmmm... I am not sure if I can confirm it, but it certainly built OK on Python 2.7.15, and selected appropriate versions of these dependencies. However, for safety, I will change them to suitable ranges.

adamjstewart commented on var/spack/repos/builtin/packages/py-jmespath/package.py:
```diff
@@ -0,0 +1,18 @@
+    """JMESPath (pronounced "james path") allows you to declaratively
+       specify how to extract elements from a JSON document"""
+
+    homepage = "git://github.com/boto/jmespath"
```
These links won't open in my browser, you should use the https version instead.

neilflood replied:
Goodness, I am not sure where those came from. Yes, https much better.

adamjstewart commented on var/spack/repos/builtin/packages/py-s3transfer/package.py:
```diff
@@ -0,0 +1,22 @@
+
+    version('0.2.0', sha256='f23d5cb7d862b104401d9021fc82e5fa0e0cf57b7660a1331425aab0c691d021')
+    version('0.1.13', sha256='90dc18e028989c609146e241ea153250be451e05ecc0c2832565231dacdf59c1')
+
```
Missing build dep on setuptools

neilflood replied:
Done

adamjstewart commented on var/spack/repos/builtin/packages/py-s3transfer/package.py:
```diff
@@ -0,0 +1,22 @@
+    version('0.2.0', sha256='f23d5cb7d862b104401d9021fc82e5fa0e0cf57b7660a1331425aab0c691d021')
+    version('0.1.13', sha256='90dc18e028989c609146e241ea153250be451e05ecc0c2832565231dacdf59c1')
+
+    depends_on('py-botocore@1.3.0:1.99', type=('build', 'run'))
```
`@1.12.36:1.99`, at least in version 0.2.0

neilflood replied:
Apologies. That came from my earlier build, using 0.1.13. Adding explicit variation of the dependency for the two versions listed.

adamjstewart commented on var/spack/repos/builtin/packages/py-s3transfer/package.py:
```diff
@@ -0,0 +1,22 @@
+    version('0.1.13', sha256='90dc18e028989c609146e241ea153250be451e05ecc0c2832565231dacdf59c1')
+
+    depends_on('py-botocore@1.3.0:1.99', type=('build', 'run'))
+    depends_on('py-futures@2.2.0:3.999', when='^python@:2.99', type=('build', 'run'))
```
You can just do `@:2` here to capture all of Python 2

neilflood replied:
Done

adamjstewart commented on var/spack/repos/builtin/packages/py-s3transfer/package.py:
```diff
@@ -0,0 +1,22 @@
+    depends_on('py-futures@2.2.0:3.999', when='^python@:2.99', type=('build', 'run'))
+    depends_on('py-nose', type='test')
+    depends_on('py-mock', type='test')
+    depends_on('py-coverage', type='test')
```
I would just drop this, it's prob not needed for testing, only for coverage reporting.

neilflood replied:
OK. It certainly works without it, so done.

neilflood commented:
@adamjstewart While checking that proj stuff, I noticed that you have been busy adding packages (jmespath, s3transfer and botocore) which I had as part of this PR for awscli. No problem, but I wondered if I should cancel this PR, since you have done most of it separately anyway, and perhaps just start a new one for the awscli package by itself. This PR seems to have been left behind. Let me know what you think.

adamjstewart commented:
I knew those packages seemed familiar, but I couldn't find them. Sorry about the duplication of effort. Yes, it would probably be easiest to create a new PR with just awscli.

neilflood commented:
this PR got left behind, and in the meantime all the dependencies were added in by others. Closing this, and will start a new PR just for awscli shortly.
